proposals/migrate-over-ipfs.md

```diff
@@ -0,0 +1,163 @@
+        // Whether or not to keep the migration after downloading it.
+        "Keep": "pin"|"cache"|"discard"
+    }
+}
+```
+
+For the initial version, the default download policy should be `["HTTPS", "IPFS"]` unless the local
+node uses private networks (has a swarm key) in which case it should just be `["HTTPS"]` for safety.
+As we gain confidence in our ability to reliably download migrations over IPFS, we should make it
+the default.
+
+When migrating a repo, until we succeed.
+
+1. If the next source is HTTPS, download from `ipfs.io`.
+2. If the source is IPFS:
+    1. Create a new _temporary_ repo (go-ipfs can't read the current repo because it's an older version).
+    2. Start a new _temporary_ go-ipfs node in the temporary repo.
+        a. This node should not listen for inbound connections as it has no way to know which ports/transports should be configured (can't read the config).
+        b. This node should not expose an API/gateway.
+    3. Download the required migration binaries using the temporary go-ipfs node.
+4. Migrate the main go-ipfs node's repo.
+5. Start the main go-ipfs node.
+6. Maybe store a copy of the migration:
+    1. If `Keep` is not "discard", import the migration binary into the main go-ipfs node's repo.
+    2. If `Keep` is "pin", pin the migration binary.
+7. If running, shutdown any temporary nodes and delete the temporary repo.
+
+**Implementation steps:**
+
+1. Implement everything but step 6.
+2. Implement step 6. Unless nodes keep around a copy of the migration, this feature isn't going to be all that useful.
+3. Eventually, switch over to using IPFS by default instead of the gateway as the first option.
+
+NOTE: the only reason we're proposing using the gateway first is that the migration blocks startup so it needs to be fast and reliable.
+
+#### What does done look like?
+_What specific deliverables should completed to consider this project done?_
+
+1. go-ipfs can download migrations over IPFS (bitswap) without having to contact a centralized server over HTTP.
+2. This feature has been tested by multiple labbers (possibly even tested by a subset of ipfs-desktop users).
+
+####  What does success look like?
+_Success means impact. How will we know we did the right thing?_
+
+Users can start a new version of go-ipfs that switches to a new repo version without downloading anything from a centralized (easy to block) server.
+
+#### Counterpoints &amp; pre-mortem
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+
+1. We may not be able to download/find the migration on IPFS. Trying the gateway first (or racing) should help mitigate this.
+2. The "temporary" node may be missing important parts of the configuration. For example, it may need alternative bootstrap nodes as our nodes may not be reachable. We'll need to think about this carefully and may need to "reach" into the old repo's config file a bit.
```

Comment:
Could totally be a problem with a few possible ways to fix it, it might not be so bad to wait for the user reports to come in before deciding how to resolve this.

One perhaps simple way to deal with this would be allowing users to pass a flag `ipfs daemon --migration-config=/path/to/config` and then load that config file into the temporary node. It's a bit of a hassle for the user, but would certainly handle the job.